homeassistant/components/bluetooth/active_update_processor.py

```diff
@@ -0,0 +1,88 @@
+"""A DataProcessor that collects data passively from advertisements but can also poll."""
+from __future__ import annotations
+
+import asyncio
+from collections.abc import Callable, Coroutine
+import time
+from typing import Any, TypeVar
+
+from homeassistant.components.bluetooth import (
+    BluetoothChange,
+    BluetoothServiceInfoBleak,
+)
+from .passive_update_processor import (
+    PassiveBluetoothDataProcessor,
+    PassiveBluetoothDataUpdate,
+)
+from homeassistant.core import HomeAssistant, callback
+
+_T = TypeVar("_T")
+
+
+class ActiveDataProcessor(PassiveBluetoothDataProcessor):
```

Comment:
Should we put the poll/update code in the top level coordinator in case we need to poll data that would be consumed by multiple processor in a single connection?

Reply:
That's a good idea.

In that case, the processor wouldn't need to know about the poll. But we would need another mechanism to trigger an update that just took a SensorUpdate.

If there was a `async_process_update` that took a `SensorUpdate` on `PassiveBluetoothDataProcessor`, then `async_handle_bluetooth_event` could call it. And then I wouldn't need a different DataProcessor subclass, just a different Coordinator.

Reply:
I'd also like to figure out a way to pass a list of `entity_key` to `needs_poll`. The idea would be if you turn off the MiFlora battery sensor, it doesn't need to poll it. It looks like the `DataProcessor` has that information via `_entity_key_listeners`. But I don't think coordinator currently does.

Reply:
If i was going to do that, `_entity_key_listeners.keys()` contains:

```
dict_keys([PassiveBluetoothEntityKey(key='illuminance', device_id=None), PassiveBluetoothEntityKey(key='battery', device_id=None), PassiveBluetoothEntityKey(key='moisture', device_id=None), PassiveBluetoothEntityKey(key='conductivity', device_id=None)])
```

It's created "just in time". So there is no battery key until we first poll the battery. So i'd have to pre-create a battery sensor in response to advertisement data.

We'd also need the opposite of `_device_key_to_bluetooth_entity_key`.

Reply:
Right now the coordinator does not really know about the sensor device class at all. This means it's hard to do both ideas at once.

If it makes sense to pull the polling up to the coordinator, maybe it would make sense to lift the parsing out of the data processor into the coordinator. E.g. for xiaomi if get an advertisement with an encrypted battery event and motion event in one packet. We don't want to parse that once in sensor and once in binary_sensor, especially as a bindkey problem might trigger a reauth event. For us, it would make more sense to have the coordinator to the parsing (and polling) and the processor just maps data between what our device class has parsed/polled and how that is represented in HA.

Reply:
The coordinator shouldn't know about entities or entity device_info. The processors handle those. This is the design separation we need to keep.

Reply:
I'm not proposing that the coordinator knows about home assistant entities. Im proposing that it drives the library that does the parsing and decoding, and only do that bit once. It just so happens that the underlying library uses similar terminology like entity for its parsed representation.

Right now the coordinator just forwards the raw binary data to each interested platform. So we have to decrypt multiple times, potentially trigger reauth from multiple places, etc.

Reply:
Good!

Reply:
When we add multiple platforms, I think we could solve the multiple reauth by registering a dedicated processor who's sole job is to check if the bind key needs to be updated and trigger reauth.  Other processors could reject the data if it can't process it knowing it the reauth processor would eventually lead to working data

Reply:
#76384 will let me pull the polling up to the coordinator (i'll be able to test the advertisement to see if its a MiFlora, then start allow polling if it is).